piecrust/src/store/tree.rs

```diff
@@ -0,0 +1,154 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+use bytecheck::CheckBytes;
+use piecrust_uplink::ModuleId;
+use rkyv::{Archive, Deserialize, Serialize};
+
+// This means we have max `2^32` modules
+const HEIGHT: usize = 32;
+const ARITY: usize = 2;
+
+pub type Tree = dusk_merkle::Tree<Hash, HEIGHT, ARITY>;
+
+#[derive(
+    Debug,
+    Clone,
+    Copy,
+    PartialEq,
+    Eq,
+    PartialOrd,
+    Ord,
+    Hash,
+    Archive,
+    Deserialize,
+    Serialize,
+    CheckBytes,
+)]
+#[archive(as = "Self")]
+pub struct Hash([u8; blake3::OUT_LEN]);
+
+impl Hash {
+    pub fn as_bytes(&self) -> &[u8; blake3::OUT_LEN] {
+        &self.0
+    }
+}
+
+impl From<Hash> for [u8; blake3::OUT_LEN] {
+    fn from(hash: Hash) -> Self {
+        hash.0
+    }
+}
+
+impl From<[u8; blake3::OUT_LEN]> for Hash {
+    fn from(bytes: [u8; blake3::OUT_LEN]) -> Self {
+        Self(bytes)
+    }
+}
+
+impl AsRef<[u8]> for Hash {
+    fn as_ref(&self) -> &[u8] {
+        &self.0
+    }
+}
+
+impl dusk_merkle::Aggregate<HEIGHT, ARITY> for Hash {
+    const EMPTY_SUBTREES: [Hash; HEIGHT] = EMPTY_SUBTREES;
+
+    fn aggregate<'a, I>(items: I) -> Self
+    where
+        Self: 'a,
+        I: Iterator<Item = &'a Self>,
+    {
+        let mut hasher = Hasher::new();
+        for item in items {
+            hasher.update(item.as_bytes());
+        }
+        hasher.finalize()
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Hasher(blake3::Hasher);
+
+impl Hasher {
+    #[inline(always)]
+    pub fn new() -> Self {
+        Self(blake3::Hasher::new())
+    }
+
+    #[inline(always)]
+    pub fn update(&mut self, input: &[u8]) -> &mut Self {
+        self.0.update(input);
+        self
+    }
+
+    #[inline(always)]
+    pub fn finalize(&self) -> Hash {
+        let hash = self.0.finalize();
+        Hash(hash.into())
+    }
+}
+
+/// Returns the position of a in the tree `module` given its ID.
+///
+/// # SAFETY:
+/// If there ever is a collision between IDs this is likely the culprit and
+/// should be changed.
```

Comment:
Maybe add a short description of how we compute the position form ModuleId.
e.g.:

To compute the position in the Module Tree, the 32-byte ModuleId is divided in 8 4-byte slices, which are summed up together (modulo 2^32) to obtain the position value.

Also, the SAFETY message should probably be removed or updated, since we prevent collisions when inserting a module. We should mention this here.

e.g.: 
SAFETY: since we are mapping 32-byte values to 4-byte values, it is possible to have collisions between different `ModuleId`s. To prevent this, we check for collisions before inserting a new module (see `deploy`).